### PR TITLE
Reproduction of and fix for #81

### PR DESCRIPTION
**Pull request by nicktindall**
See the issue for details on the problem.

This should be functionally equivalent to the old way, except it will guarantee mutual exclusion in the presence of a lock.

### CODE REVIEW

File: affinity/src/main/java/net/openhft/affinity/LockCheck.java [hunk 1]
```diff
@@ -109,15 +109,10 @@ static int getProcessForCpu(int core) throws IOException {
...
+    static void updateCpu(int cpu) throws IOException {
         if (!canOSSupportOperation())
             return;
-        try {
-            replacePid(cpu, getPID());
-        } catch (IOException e) {
-            LOGGER.warn("Failed to update lock file for cpu " + cpu, e);
-            e.printStackTrace();
-        }
+        replacePid(cpu, getPID());
```
Comment (nicktindall) [inline affinity/src/main/java/net/openhft/affinity/LockCheck.java]: stop swallowing the exception

File: affinity/src/main/java/net/openhft/affinity/LockInventory.java [hunk 2]
```diff
@@ -69,9 +70,22 @@ private static boolean isAnyCpu(final int cpuId) {
...
+     * @return true if the lock was acquired, false otherwise
+     */
+    private static boolean updateLockForCurrentThread(final boolean bind, final AffinityLock al, final boolean wholeCore) {
+        try {
+            LockCheck.updateCpu(al.cpuId());
+            al.assignCurrentThread(bind, wholeCore);
+            return true;
+        } catch (IOException e) {
+            return false;
+        }
```
Comment (nicktindall) [inline affinity/src/main/java/net/openhft/affinity/LockInventory.java:89]: return a boolean to indicate whether the lock was acquired

File: affinity/src/main/java/net/openhft/affinity/LockInventory.java [hunk 3]
```diff
@@ -106,8 +120,9 @@ public final synchronized AffinityLock acquireLock(boolean bind, int cpuId, Affi
         final boolean specificCpuRequested = !isAnyCpu(cpuId);
         if (specificCpuRequested && cpuId != 0) {
             final AffinityLock required = logicalCoreLocks[cpuId];
-            if (required.canReserve(true) && anyStrategyMatches(cpuId, cpuId, strategies)) {
-                updateLockForCurrentThread(bind, required, false);
+            if (required.canReserve(true)
+                    && anyStrategyMatches(cpuId, cpuId, strategies)
+                    && updateLockForCurrentThread(bind, required, false)) {
                 return required;
             }
```
Comment (nicktindall) [inline affinity/src/main/java/net/openhft/affinity/LockInventory.java:129]: only return the lock if it was successfully acquired

File: affinity/src/main/java/net/openhft/affinity/LockInventory.java [hunk 4]
```diff
@@ -119,8 +134,9 @@ public final synchronized AffinityLock acquireLock(boolean bind, int cpuId, Affi
             // if you have only one core, this library is not appropriate in any case.
             for (int i = logicalCoreLocks.length - 1; i > 0; i--) {
                 AffinityLock al = logicalCoreLocks[i];
-                if (al.canReserve(false) && (isAnyCpu(cpuId) || strategy.matches(cpuId, al.cpuId()))) {
-                    updateLockForCurrentThread(bind, al, false);
+                if (al.canReserve(false)
+                        && (isAnyCpu(cpuId) || strategy.matches(cpuId, al.cpuId()))
+                        && updateLockForCurrentThread(bind, al, false)) {
                     return al;
```
Comment (nicktindall) [inline affinity/src/main/java/net/openhft/affinity/LockInventory.java:142]: only return the lock if it was successfully acquired

File: affinity/src/main/java/net/openhft/affinity/LockInventory.java [hunk 5]
```diff
@@ -136,8 +152,8 @@ public final synchronized AffinityLock tryAcquireLock(boolean bind, int cpuId) {
             return null;
 
         final AffinityLock required = logicalCoreLocks[cpuId];
-        if (required.canReserve(true)) {
-            updateLockForCurrentThread(bind, required, false);
+        if (required.canReserve(true)
+                && updateLockForCurrentThread(bind, required, false)) {
             return required;
```
Comment (nicktindall) [inline affinity/src/main/java/net/openhft/affinity/LockInventory.java:159]: only return the lock if it was successfully acquired

File: affinity/src/main/java/net/openhft/affinity/LockInventory.java [hunk 6]
```diff
@@ -156,8 +172,9 @@ public final synchronized AffinityLock acquireCore(boolean bind, int cpuId, Affi
                         continue LOOP;
 
                 final AffinityLock al = als[0];
-                updateLockForCurrentThread(bind, al, true);
-                return al;
+                if (updateLockForCurrentThread(bind, al, true)) {
+                    return al;
+                }
```
Comment (nicktindall) [inline affinity/src/main/java/net/openhft/affinity/LockInventory.java:179]: only return the lock if it was successfully acquired

File: affinity/src/main/java/net/openhft/affinity/lockchecker/FileLockBasedLockChecker.java [hunk 7]
```diff
@@ -61,22 +63,21 @@ private boolean isLockFree(File file, int id) {
...
+            try (FileLock fileLock = fc.tryLock()) {
+                if (fileLock == null) {
+                    return false;
+                }
             }
-        } catch (IOException | OverlappingFileLockException e) {
+            // file is present but nobody had it locked, the process holding it must have
+            // failed. Don't try and delete it because it can cause a race with other
+            // processes attempting to lock it.
+            return true;
```
Comment (nicktindall) [inline affinity/src/main/java/net/openhft/affinity/lockchecker/FileLockBasedLockChecker.java]: don't delete lock files left by processes that died, just act as though they weren't there

File: affinity/src/main/java/net/openhft/affinity/lockchecker/FileBasedLockChecker.java [hunk 8]
```diff
@@ -1,83 +0,0 @@
...
-import org.slf4j.LoggerFactory;
-
-import java.io.*;
-import java.text.SimpleDateFormat;
-import java.util.Date;
-
-import static java.nio.charset.StandardCharsets.UTF_8;
-
-@SuppressWarnings("ResultOfMethodCallIgnored")
-public class FileBasedLockChecker implements LockChecker {
```
Comment (JerryShea) [inline affinity/src/main/java/net/openhft/affinity/lockchecker/FileBasedLockChecker.java:14]: shouldn't we deprecate if part of public API?

File: affinity/src/main/java/net/openhft/affinity/lockchecker/FileLockBasedLockChecker.java (hunk 7 above)
Comment (JerryShea) [reply affinity/src/main/java/net/openhft/affinity/lockchecker/FileLockBasedLockChecker.java]: so this is the only significant behaviour change? The rest of this PR is uncontroversial bug fixes?

Comment (nicktindall) [reply affinity/src/main/java/net/openhft/affinity/lockchecker/FileLockBasedLockChecker.java]: Yeah the lock implementation will no longer aggressively clean up after badly behaved processes that don't release their locks

**Review by JerryShea — APPROVED** [review]

File: affinity/src/main/java/net/openhft/affinity/lockchecker/FileBasedLockChecker.java (hunk 8 above)
Comment (nicktindall) [reply affinity/src/main/java/net/openhft/affinity/lockchecker/FileBasedLockChecker.java:14]: Yeah possibly, although it's not something that would be possible use, the lock checker implementation is baked in as a private static final field (see https://github.com/OpenHFT/Java-Thread-Affinity/blob/0bdef1808617b84d9391a7859baf024f16457a36/affinity/src/main/java/net/openhft/affinity/LockCheck.java#L39) so it seems safe to just drop it?

File: affinity/src/test/java/net/openhft/affinity/FileLockLockCheckTest.java [hunk 9]
```diff
@@ -53,6 +53,7 @@ public void before() {
 
     @After
     public void after() {
+        LockCheck.releaseLock(cpu);
```
Comment (nicktindall) [inline affinity/src/test/java/net/openhft/affinity/FileLockLockCheckTest.java]: These tests used to leave state in the static locks collection in FileLockLockCheck

File: affinity/src/main/java/net/openhft/affinity/lockchecker/FileLockBasedLockChecker.java (hunk 7 above)
Comment (nicktindall) [reply affinity/src/main/java/net/openhft/affinity/lockchecker/FileLockBasedLockChecker.java]: Update, the lock files will never be deleted it seemed as though that caused race conditions across processes.

**Review by rogersimmons — COMMENTED** [review]
If the intention is to guard exclusive access to a cpu via a lock file, the existence of the file itself is enough to indicate the cpu is locked. "Create a file if it doesn't exist already, or return error" is an atomic operation and can be used for this type of "mutex" without resorting to actual file locks. (Avoiding the problem that you can't atomically open-and-lock a file). Once the file is open then details can be written to it as needed since you know you are the only user (assuming cooperating processes). Or for belt and braces you can take a lock at that point too.

A process failing without releasing the lock can be handled by checking the content of the file (opening normally in a cooperative "unlocked" fashion to check contents after failing the atomic-create-if-not-exists) and eg checking the pid inside etc, then forcing things as needed)

**Comment by nicktindall** [comment]
That's an interesting point. I wonder why the file lock version was implemented now. There was a non lock version as well but it wasn't used.

**Comment by minborg** [comment]
I think it would make sense to have a library that provides inter-process locking of various sorts. In Map, for example, we would need a lock that will release when the holding process dies. So, in this case, we might lock that portion of the file and CAS to acquire. If the lock remains acquired for a longer period of time, we can check if the file lock is still there. If it's not, then the original process died and we may safely assume the lock is implicitly released and can be re-acquired by another process. We could then get rid of unsafe "wait-for-15-seconds-and-then-just-grab-the-lock" solutions and we can determine process deaths much faster. So if a process dies, we can potentially be back in 10 ms rather than 10 seconds.

**Comment by nicktindall** [comment]
I updated the PR to not use locks, the process looks like

to acquire
```
1: if (lock file exists) and (lock file contains a live PID)
2:    fail to acquire
3: delete lock file (to cover the dead-PID case)
4: open file channel, creating lock file or fail if it already exists (atomic)
5: write PID to it
```

to release
```
1: close file channel
2: delete file
```

It doesn't seem like it should be atomic to me, but the tests indicate it is. The issue I thought would exist is when a orphaned lock was left and P1 and P2 attempt to acquire (using line numbers above)....
```
P2: 1
P1: 1
P1: 3
P1: 4
P2: 3 <--- the delete must fail here because P1 has the file open? 
```

But it seems to work. Or at least works a lot better than what was there (and is more tested)

**Comment by nicktindall** [comment]
> I think it would make sense to have a library that provides inter-process locking of various sorts. In Map, for example, we would need a lock that will release when the holding process dies. So, in this case, we might lock that portion of the file and CAS to acquire. If the lock remains acquired for a longer period of time, we can check if the file lock is still there. If it's not, then the original process died and we may safely assume the lock is implicitly released and can be re-acquired by another process. We could then get rid of unsafe "wait-for-15-seconds-and-then-just-grab-the-lock" solutions and we can determine process deaths much faster. So if a process dies, we can potentially be back in 10 ms rather than 10 seconds.

I agree this would be useful, but for this particular case I think the simplest thing should be used because this library is very low level.

**Comment by minborg** [comment]
The PID may appear to be sufficiently random for this solution to work reasonably. However, if a process is running under Docker, it is likely that all instances will have the same PID (e.g. 5).  In the future, we should consider having more unique data in the lock key.

**Review by rogersimmons — CHANGES_REQUESTED** [review]
Hi Nick - there is indeed a race condition where you suspected: the delete (P2: 3) is not safe (tested on Linux).

The following strategy fixes this, and also addresses Per's issue re Docker & PIDs.
The tweak is to create and lock a file off to one side, before trying to (atomically) move it in to the shared position.

1. create a temporary file (eg by Linux mktemp)
2. lock the file
3. atomically rename to the shared lock file eg "cpuN.lock" (eg using Linux renameat2 with RENAME_NOREPLACE). This will either fail if the file already exists, or perform the equivalent of an atomic-create-and-lock-file
4. if succeeded you now have the lock. If failed, see 3a-d below
5. ... use 
6. delete the lock file when done

If step 3 fails you either legitimately don't have the lock, or a process with the lock died without correctly cleaning up.
If the latter, the OS will have released the lock, allowing the following to safely inherit where appropriate (note no PID check or timeout/forcing):
3a. open the lock file normally for r/w. if this fails go back to step 1
3b. try to acquire the lock
3c. if the lock acquisition (3b) worked then the old process detached incorrectly. You have inherited the lock. Clean up any tmp files. Use the shared resource. delete the lock file when done
3d. if the lock acquisition (3b) was denied then the lock is legitimately held by someone else (either normally or by inheritance)

**Comment by nicktindall** [comment]
> Hi Nick - there is indeed a race condition where you suspected: the delete (P2: 3) is not safe (tested on Linux).
> 
> The following strategy fixes this, and also addresses Per's issue re Docker & PIDs. The tweak is to create and lock a file off to one side, before trying to (atomically) move it in to the shared position.
> 
> 1. create a temporary file (eg by Linux mktemp)
> 2. lock the file
> 3. atomically rename to the shared lock file eg "cpuN.lock" (eg using Linux renameat2 with RENAME_NOREPLACE). This will either fail if the file already exists, or perform the equivalent of an atomic-create-and-lock-file
> 4. if succeeded you now have the lock. If failed, see 3a-d below
> 5. ... use
> 6. delete the lock file when done
> 
> If step 3 fails you either legitimately don't have the lock, or a process with the lock died without correctly cleaning up. If the latter, the OS will have released the lock, allowing the following to safely inherit where appropriate (note no PID check or timeout/forcing): 3a. open the lock file normally for r/w. if this fails go back to step 1 3b. try to acquire the lock 3c. if the lock acquisition (3b) worked then the old process detached incorrectly. You have inherited the lock. Clean up any tmp files. Use the shared resource. delete the lock file when done 3d. if the lock acquisition (3b) was denied then the lock is legitimately held by someone else (either normally or by inheritance)

I think there's still a race condition there
```
P1: 1, 2, 3, 4, 5
P2: 1, 2, 3, 3a
P1: 6
P2: 3b,  3c... (it has a lock on a non-existent file)
P3: 1, 2, 3, 4, 5 (it's able to acquire the lock because there was no lockfile)
```

I think this might work for `acquireLock()`
1. Create new or open existing lock file (covers the case where a process doesn't clean up)
2. Acquire exclusive lock on lock file
3. If acquisition fails another process has the lock, return false
4. If acquisition succeeds, check that the lock file exists, if it doesn't someone else has deleted it, go to 1 (covers off the case where the file was deleted between open and acquire lock)
5. Write metadata to lock file, do work
5. Delete the lock file
6. Release the lock, close the channel

This also means we can clean up orphans on `isFree()`
1. Open existing lock file, if it doesn't exist return true
2. Acquire shared lock on the lock file. if that fails, someone else has an exclusive lock, return true
3. If acquire succeeds, it was an orphaned file, delete it, return true
4. Release lock

I _think_ mutual exclusion is guaranteed because
1. Nobody deletes without first having a lock
2. After the locker acquires the lock, the first thing they do is check that the file exists and release the lock if it doesn't.

File: affinity/src/main/java/net/openhft/affinity/LockCheck.java [hunk 10]
```diff
@@ -109,15 +91,10 @@ static int getProcessForCpu(int core) throws IOException {
...
         if (!canOSSupportOperation())
-            return;
-        try {
-            replacePid(cpu, getPID());
-        } catch (IOException e) {
-            LOGGER.warn("Failed to update lock file for cpu " + cpu, e);
-            e.printStackTrace();
-        }
+            return true;
+        return replacePid(cpu, getPID());
```
Comment (nicktindall) [inline affinity/src/main/java/net/openhft/affinity/LockCheck.java:97]: Indicate whether acquire was successful

File: affinity/src/main/java/net/openhft/affinity/LockCheck.java [hunk 11]
```diff
@@ -86,17 +70,15 @@ static boolean isProcessRunning(long pid) {
      * stores the pid in a file, named by the core, the pid is written to the file with the date
      * below
      */
-    private synchronized static void storePid(long processID, int cpu) throws IOException {
-        if (!lockChecker.obtainLock(cpu, Long.toString(processID))) {
-            throw new IOException(String.format("Cannot obtain file lock for cpu %d", cpu));
-        }
+    private synchronized static boolean storePid(long processID, int cpu) throws IOException {
+        return lockChecker.obtainLock(cpu, Long.toString(processID));
```
Comment (nicktindall) [inline affinity/src/main/java/net/openhft/affinity/LockCheck.java:74]: indicate whether acquire was successful

File: affinity/src/main/java/net/openhft/affinity/LockCheck.java [hunk 12]
```diff
@@ -52,30 +52,14 @@ public static boolean isCpuFree(int cpu) {
...
-            }
-            return false;
-        }
+        return isLockFree(cpu);
     }
 
-    static void replacePid(int cpu, long processID) throws IOException {
-        storePid(processID, cpu);
+    static boolean replacePid(int cpu, long processID) throws IOException {
+        return storePid(processID, cpu);
```
Comment (nicktindall) [inline affinity/src/main/java/net/openhft/affinity/LockCheck.java:59]: indicate whether it was successful

**Comment by rogersimmons** [comment]
Thanks Nick - looks good

**Review by rogersimmons — APPROVED** [review]

**Pull request merged**